src/pycapsule.rs

```diff
@@ -0,0 +1,417 @@
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use pyo3::{prelude::*, pycapsule::PyCapsule};
+    /// use std::ffi::CString;
+    ///
+    /// Python::with_gil(|py| {
+    ///     let name = CString::new("foo").unwrap();
+    ///     let capsule = PyCapsule::new(py, 123_u32, &name).unwrap();
+    ///     let val = unsafe { capsule.reference::<u32>() };
+    ///     assert_eq!(*val, 123);
+    /// });
+    /// ```
+    ///
+    /// However, attempting to construct a `PyCapsule` with a zero sized type will not compile:
+    ///
+    /// ```compile_fail
+    /// use pyo3::{prelude::*, pycapsule::PyCapsule};
+    /// use std::ffi::CString;
+    ///
+    /// Python::with_gil(|py| {
+    ///     let name = CString::new("foo").unwrap();
+    ///     let capsule = PyCapsule::new(py, (), &name).unwrap();  // Oops! `()` is zero sized!
+    /// });
+    /// ```
+    pub fn new<'py, T: 'static + Send>(
+        py: Python<'py>,
+        value: T,
+        name: &CStr,
+    ) -> PyResult<&'py Self> {
+        Self::new_with_destructor(py, value, name, |_, _| {})
+    }
+
+    /// Constructs a new capsule whose contents are `value`, associated with `name`.
+    ///
+    /// Also provides a destructor: when the `PyCapsule` is destroyed, it will be passed the original object,
+    /// as well as `*mut c_void` which will point to the capsule's context, if any.
+    ///
+    /// # Panics
+    ///
+    /// An attempt to add a zero sized value will panic.
+    pub fn new_with_destructor<
+        'py,
+        T: 'static + Send + PanicWhenZeroSized,
+        F: FnOnce(T, *mut c_void),
+    >(
+        py: Python<'py>,
+        value: T,
+        name: &CStr,
+        destructor: F,
+    ) -> PyResult<&'py Self> {
+        PanicWhenZeroSized::assert_not_zero_sized(&value);
+        let val = Box::new(CapsuleContents { value, destructor });
+
+        let cap_ptr = unsafe {
+            ffi::PyCapsule_New(
+                Box::into_raw(val) as *mut c_void,
+                name.as_ptr(),
+                Some(capsule_destructor::<T, F>),
+            )
+        };
+        unsafe { py.from_owned_ptr_or_err(cap_ptr) }
+    }
+
+    /// Imports an existing capsule.
+    ///
+    /// The `name` should match the path to the module attribute exactly in the form
+    /// of `module.attribute`, which should be the same as the name within the capsule.
+    ///
+    /// # Safety
+    ///
+    /// It must be known that this capsule's pointer is to an item of type `T`.
+    pub unsafe fn import<'py, T>(py: Python<'py>, name: &CStr) -> PyResult<&'py T> {
+        let ptr = ffi::PyCapsule_Import(name.as_ptr(), false as c_int);
+        if ptr.is_null() {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(&*(ptr as *const T))
+        }
+    }
+
+    /// Sets the context pointer in the capsule.
+    ///
+    /// # Notes
+    ///
+    /// The default destructor does not drop the context value. If you need to ensure it is not
```

Comment:
These two paragraphs (this one and "Finally...") are now out of date again :( Suggestion:

```
If you want to store a Rust value as the context, and drop it from the destructor, use `Box::into_raw` to convert it into a pointer, see the example.
```